standards/nestjs_standards.md

```diff
@@ -0,0 +1,372 @@
+│   │   ├── users.controller.ts
+│   │   ├── users.module.ts
+│   │   ├── users.service.ts
+│   │   └── users.spec.ts
+├── common/              # Shared utilities
+│   ├── filters/         # Exception filters
+│   ├── interceptors/    # Request/response interceptors
+│   ├── guards/          # Auth guards
+│   ├── pipes/           # Validation pipes
+│   └── decorators/      # Custom decorators
+├── config/              # Configuration files
+├── main.ts              # Application entry point
+└── app.module.ts        # Root module
+```
+
+### Module Organization Principles
+
+- Each feature in its own module; cross-cutting concerns (auth, email, metrics) go in common or shared modules.
+- Small apps: flat structure; larger apps: domain or hexagonal architecture per module.
+- Use barrel files for clean imports—avoid hidden dependencies.
+- Use `forwardRef()` judiciously to resolve circular references.
+
+## Coding Standards
+
+### General Guidelines
+
+```ts
+// Example: Async/Await
+async function fetchData() {
+  const result = await this.httpService
+    .get("https://api.example.com")
+    .toPromise();
+  return result.data;
+}
+```
+
+- Follow **TypeScript** conventions and leverage strong typing
+- Prefer `async/await` over Promises directly
+- Use **ESLint** and **Prettier** for consistent style enforcement
+- Avoid magic strings/numbers — use enums or constants
+
+```ts
+// Example: Constants
+export const DEFAULT_ROLE = "user";
+```
+
+- Name files consistently (`.controller.ts`, `.service.ts`, etc.)
+
+### Naming Conventions
+
+- `camelCase` for variables and functions
+- `PascalCase` for classes and interfaces
+- `snake_case` for database columns (if applicable)
+- `UPPER_CASE` for environment variables
+
+### Service & Controller Patterns
+
+```ts
+// users.controller.ts
+@Controller("users")
+export class UsersController {
+  constructor(private readonly usersService: UsersService) {}
+
+  @Get()
+  getAllUsers() {
+    return this.usersService.findAll();
+  }
+}
+
+// users.service.ts
+@Injectable()
+export class UsersService {
+  findAll() {
+    return [{ id: 1, name: "Alice" }];
+  }
+}
+```
+
+- Keep controllers **thin**; delegate business logic to services
+- Services should be **stateless** and reusable
+- Do not inject repositories directly into controllers
+
+### DTOs (Data Transfer Objects) & Validation
+
+```ts
+// create-user.dto.ts
+import { IsEmail, IsNotEmpty } from "class-validator";
+
+export class CreateUserDto {
+  @IsEmail()
+  email: string;
+
+  @IsNotEmpty()
+  password: string;
+}
+
+// main.ts
+app.useGlobalPipes(new ValidationPipe());
+```
+
+## Guards
+
+Guards determine whether a request should be handled by the route handler.
+
+```ts
+// roles.guard.ts
+@Injectable()
+export class RolesGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean {
+    const request = context.switchToHttp().getRequest();
+    const user = request.user;
+    return user?.roles?.includes('admin');
```

Comment:
```suggestion
    return user?.roles?.includes('admin');
```

Roles should be enum, right?
Also I think we're permission-based instead of role-based